feat: Dashboard Level Access : dashboard lists

Ofeknielsen opened the pull request:
## Summary 

As the first phase of [SIP-51 "Proposal for Dashboard Level Access"](https://github.com/apache/incubator-superset/issues/10408), added dashboard permission enforcement for getting dashboards list (via superset API or view).
The feature is activated by turn the flag *"DASHBOARD_LEVEL_ACCESS"* on.
When request to get dashboard list is handled, the *"dashboard filter"* filter the dashboards by the user dashboard access permissions, indeed answer the question does the user have "all dashboards permissions" or permissions for specific dashboards.
As default behavior, Gamma users have all dashboards access permission
On create/edit/copy dashboards - dashboards view menu and permissionView are created/updated.
On delete, the dashboards view menu and permissionView are deleted
On superset init phase (while the flag is on) - permissions created for all the exists dashboards.

## TEST PLAN
Added new github action: Feature Flag - Dashboard Level Access 
Using "pytest skip-if" on tests which tests previous logic (when the flag is off) 

## ADDITIONAL INFORMATION
 
With the feature, we have tried to increase the quality of the code by meeting the standard programming principles:
### Don't Repeat Yourself principle
1. Use constants instead of reusing "hard coded"
2. Delete filters from superset.view.dashboard
3. DashboardSecurityManager instead writing duplicate code in security-manager ("can access all")
4. Dashboard utils inside tests.dashboard
### Separation Of Concerns and Single Responsibility principles 
1. Added dashbaord_level_access_initializer.py - instead writing the logic inside the cli file.
2. DashboardSecurityMixin instead writing the code in dashboard model or security_manager
3. DashboardSecurityManager instead writing the code in security_manager
4. Refactor set_perm by extracting code to set_permissions_views 
5. Writing update_view_menu and find_view_menu_by_pattern as general code in security_manager instead in the callee files.
### Tests - dashboard test refactor
1. Decouple tests from specific data - 
2. Decouple tests to other tests 
### Python best practises 
1. use of "__" in method/variables (__set_permission_by_connection in security_manager)

resolve #10408

Commit:
Add all dashboards permissions views

Commit:
Add all dashboards permissions views

Commit:
Fix format

Commit:
Fix format

amitNielsen commented:
https://github.com/apache/incubator-superset/issues/10408

Commit:
Fix format

Commit:
Fix format

Commit:
Add dashboard permission view on new dashboard

Commit:
Add dashboard permission view on new dashboard

Commit:
feat update dashboard permissions on update dashboard title

feat delete dashboard permission on delete dashboard

Commit:
feat update dashboard permissions on update dashboard title

feat delete dashboard permission on delete dashboard

Commit:
Merge branch 'master' into feat/add_all_dashboards_permissions

# Conflicts:
#	superset/models/dashboard.py

Commit:
Fix tests after merge from upstream/master

Commit:
Fix tests after merge from upstream/master

amitmiran137 committed:
fix: comment after_insert 

test are failing bc of the after insert that prevents fetching the dashboard with FK relationships

amitmiran137 committed:
fix: add apache license

amitmiran137 committed:
fix: readd after_insert

amitNielsen commented on superset/models/dashboard.py:
```diff
@@ -496,3 +548,13 @@ def event_after_dashboard_changed(  # pylint: disable=unused-argument
     sqla.event.listen(Dashboard, "after_update", event_after_dashboard_changed)
+
+
+sqla.event.listen(Dashboard, "after_insert", SecuredMixin.after_insert)
```
@villebro @dpgaspar @altef 
Hey,

we wanted to create the matching permission when a dashboard gets created so we did the same as was done for `datasource`

we are experiencing an issue where after adding the **after_insert** even listening some existing tests are failing (different ones on different databases)

it seems like some of the parameters like (created_by) are not being fetched from the databases after a dashboard is created

if we debug the test and adding a manual `session.commit()` after dashboard creation and before fetching the dashboard again so only then the created_by gets updates

Do you have any idea how should we address this?
is there another strategy that we should take when implementing such a behavior?

Ofeknielsen replied:
some more details

amitmiran137 committed:
fix: pylint

aspedrosa commented:
Closes #10413
Closes #10564

aspedrosa commented:
Resolve #10413, resolve #10564

Commit:
Add all dashboard access permission enforcement

Commit:
Merge remote-tracking branch 'origin/feat/add_all_dashboards_permissions' into feat/add_all_dashboards_permissions

# Conflicts:
#	superset/models/dashboard.py
#	superset/security/manager.py
#	tests/dashboards/utils.py
#	tests/security_tests.py

Commit:
Add dashboard access permission enforcement on get dashboard list

amitNielsen commented on tests/dashboard_tests.py:
```diff
@@ -465,6 +465,80 @@ def test_users_can_view_published_dashboard(self):
+
+        # Create a published and hidden dashboard and add them to the database
+        first_dash = Dashboard()
+        first_dash.dashboard_title = "Published Dashboard"
```
first dashboard

amitNielsen commented on tests/dashboard_tests.py:
```diff
@@ -465,6 +465,80 @@ def test_users_can_view_published_dashboard(self):
+        first_dash.published = True
+
+        second_dash = Dashboard()
+        second_dash.dashboard_title = "Hidden Dashboard"
```
second dashboard

amitNielsen commented on tests/dashboard_tests.py:
```diff
@@ -465,6 +465,80 @@ def test_users_can_view_published_dashboard(self):
+        second_dash_slug = f"second_dash_{random()}"
+
+        # Create a published and hidden dashboard and add them to the database
+        first_dash = Dashboard()
+        first_dash.dashboard_title = "Published Dashboard"
+        first_dash.slug = first_dash_slug
+        first_dash.slices = [slice]
+        first_dash.published = True
```
extract to dashboard_tests_utils.create_dashboard()

amitNielsen commented on tests/dashboard_tests.py:
```diff
@@ -465,6 +465,80 @@ def test_users_can_view_published_dashboard(self):
+        self.grant_access_to_dashboard(second_dash)
+
+        try:
+            db.session.merge(first_dash)
+            db.session.merge(second_dash)
```
extract to dashboard_test_utlis.delete_dashboard()

amitNielsen commented on tests/dashboard_tests.py:
```diff
@@ -465,6 +465,80 @@ def test_users_can_view_published_dashboard(self):
+            self.assertIn(f"/superset/dashboard/{second_dash_slug}/", resp)
+            self.assertIn(f"/superset/dashboard/{first_dash_slug}/", resp)
+        finally:
+            db.session.delete(first_dash)
+            db.session.delete(second_dash)
```
same

amitNielsen commented on tests/dashboard_tests.py:
```diff
@@ -465,6 +465,80 @@ def test_users_can_view_published_dashboard(self):
+            db.session.merge(second_dash)
+            db.session.commit()
+
+            resp = self.get_resp("/api/v1/dashboard/")
```
extract to dashboard_test_utlis.get_dashboards()

amitNielsen commented on tests/dashboard_tests.py:
```diff
@@ -465,6 +465,80 @@ def test_users_can_view_published_dashboard(self):
+            db.session.delete(first_dash)
+            db.session.delete(second_dash)
+            db.session.commit()
+            self.revoke_public_access_to_table(table)
```
re-use  dashboard_test_utlis.*

amitNielsen commented on superset/constants.py:
```diff
@@ -62,3 +62,14 @@ class RouteMethod:  # pylint: disable=too-few-public-methods
+class Security:  # pylint: disable=too-few-public-methods
+    class AllDashboard:
+        VIEW_NAME = "all_dashboards"
+        ACCESS_PERMISSION_NAME = "can_access_dashboard"
```
ACCESS_PERMISSION_KEY

amitNielsen commented on tests/dashboards/utils.py:
```diff
@@ -0,0 +1,37 @@
+from superset import security_manager
+
+
```
this should be DashboardTestUtilsClass where you not only have all existing methods you also manage local state of created dashboards and dashboardRoles and permissions so you can clean after every test/alltests

amitNielsen commented on superset/security/manager.py:
```diff
@@ -985,3 +1074,21 @@ def get_rls_ids(self, table: "BaseDatasource") -> List[int]:
         ids.sort()  # Combinations rather than permutations
         return ids
+
+    def del_role(self, name: str) -> bool:
```
should this method be on security manager?

amitmiran137 reviewed (commented):
Consider also Extract permission_view responsability into a new class : PermissionViewSecurityManager
same as you did for dashboards

Commit:
Removed duplicated file and migrate to use only one dashboard filter

Commit:
Refactor dashboards test to be more understandable

Commit:
Refactor dashboards test to be more understandable

Commit:
Refactor dashboards test to be more understandable

Commit:
Merge remote-tracking branch 'upstream/master' into feat/all_dashboard_permissions

# Conflicts:
#	superset/security/manager.py
#	tests/dashboard_tests.py

Commit:
Fix using wrong of copy dashboard lead to other tests to fail

Commit:
fix: per-commit

Commit:
fix: license

Commit:
fix: pre-commit

Commit:
Add feature_flag to all dashboards permissions views

Commit:
fix: lint by disable

Commit:
fix: pre commit

Commit:
Revert "Add feature_flag to all dashboards permissions views"

This reverts commit 6b755952

Commit:
fix: pre commit

Commit:
fix : lint

Commit:
fix : try to fix lint again :)

Commit:
fix : try to fix lint again :) :)

Commit:
Fix bad use dashboard level access feature flag

Commit:
fix: pre-commit

amitNielsen commented on tests/base_tests.py:
```diff
@@ -250,6 +250,52 @@ def grant_public_access_to_table(self, table):
+        pvm_pairs = [(used_consts.ACCESS_PERMISSION_NAME, view_name)]
+        if edit_too and not dashboard:
+            pvm_pairs.append((used_consts.EDIT_PERMISSION_NAME, view_name,))
+        return map(
+            lambda pvm: security_manager.find_permission_view_menu(pvm[0], pvm[1]),
+            pvm_pairs,
+        )
+
+    def grant_access_to_all_dashboards(
+        self, is_dashboard_level_access_enabled=True, role_name="Public"
+    ):
+        self.grant_access_to_dashboard(
+            None, is_dashboard_level_access_enabled, role_name
+        )
+
+    def revoke_access_to_all_dashboards(
+        self, is_dashboard_level_access_enabled=True, role_name="Public"
+    ):
+        self.revoke_access_to_dashboard(
+            None, is_dashboard_level_access_enabled, role_name
+        )
+
+    def grant_access_to_dashboard(
+        self, dashboard, is_dashboard_level_access_enabled=True, role_name="Public"
+    ):
+        if is_dashboard_level_access_enabled:
+            role = security_manager.find_role(role_name)
+            pvs = self.get_dashboards_access_permission_views(dashboard)
+            for pv in pvs:
+                security_manager.add_permission_role(role, pv)
+
+    def revoke_access_to_dashboard(
+        self, dashboard, is_dashboard_level_access_enabled=True, role_name="Public"
+    ):
+        if is_dashboard_level_access_enabled:
+            role = security_manager.find_role(role_name)
+            pvs = self.get_dashboards_access_permission_views(dashboard)
+            for pv in pvs:
+                security_manager.del_permission_role(role, pv)
+
```
this should be extracted to different test utils per a model(dashboard,chart, permissionsView)
otherwise base_tests will become huge and overtime not maintainable

Commit:
Add feature_flag to dashboard filter and tests

amitNielsen commented:
resolve #10749

Commit:
Add migration for dashboard_level_access

Commit:
Add test config for dashboard level access

Commit:
Refactor move migration for dashboard_level_access to superset init phase

Commit:
fix: pre-commit

amitmiran137 committed:
chore: log feature flag status

Commit:
fix: wrong syntax and pre-commit

Commit:
Merge remote-tracking branch 'upstream/master' into feat/add_all_dashboards_permissions

* upstream/master: (32 commits)
  docs: Add a note to contributing.md on reporting security vulnerabilities (#10796)
  Fix: Include RLS filters for cache keys (#10805)
  feat: filters for database list view (#10772)
  fix: MVC show saved query (#10781)
  added creator column and adjusted order columns (#10789)
  security: disallow uuid package on jinja2 (#10794)
  feat: CRUD REST API for saved queries (#10777)
  fix: disable domain sharding on explore view (#10787)
  fix: can not type `0.05` in `TextControl` (#10778)
  fix: pivot table timestamp grouping (#10774)
  fix: add validator information to email/slack alerts (#10762)
  More Label touchups (margins) (#10722)
  fix: dashboard extra filters (#10692)
  fix: re-installing local superset in cache image (#10766)
  feat: SIP-34 table list view for databases (#10705)
  refactor: convert DatasetList schema filter to use new distinct api (#10746)
  chore: removing fsevents dependency (#10751)
  Fix precommit hook for docs/installation.rst (#10759)
  feat(database): POST, PUT, DELETE API endpoints (#10741)
  docs: Update OAuth configuration in installation.rst (#10748)
  ...

Commit:
chore:  renaming and delete thee if is_dashboard_level_access_enabled

    sqla.event.listen

Commit:
fix: create dashboard tests are working

Commit:
fix: make more tests work
feature: bulk delete work

Commit:
fix: license

Commit:
fix: few pre-commits

Commit:
fix: prevent cyclic import using PEP 563 -- Postponed Evaluation of Annotations https://www.python.org/dev/peps/pep-0563/

Commit:
fix: delete dash arrange parse created dashboard

Commit:
Refactor add and delete permissions for dashboards

amitNielsen commented on superset/security/manager.py:
```diff
@@ -967,26 +947,20 @@ def __find_permission_view_menu_by_instances(  # pylint: disable=C0103
         """
-        results = self.get_session.query(self.viewmenu_model).filter(
-            ViewMenu.name.like(f"dashboard.%")
+        pattern_filter = (
```
nice one

amitNielsen commented on tests/dashboard_tests.py:
```diff
@@ -683,9 +684,17 @@ def test_get_dashboards__public_user_get_published(self):
-            db.session.merge(dashboard_not_to_access)
-            db.session.commit()
+            do_commit = False
+            if published_value_of_not_accessed_dashboard:
+                dashboard_not_to_access.published = True
+                db.session.merge(dashboard_not_to_access)
+                do_commit = True
+            if not published_value_of_accessed_dashboard:
+                dashboard_to_access.published = False
+                db.session.merge(dashboard_to_access)
+                do_commit = True
```
dup code

Commit:
Fix lint

Commit:
Fix failed test

Commit:
Fix failed test

Commit:
add positional arguments to pytest

amitmiran137 committed:
Merge branch 'master' into feat/add_all_dashboards_permissions

Commit:
fix: as long as that database has dashboards test is valid
this was changed bc dashboard count might change bc of deletion or permission and it is better not to be coupled to a hardcoded count

Commit:
fix: no need to clean new users in dashboard api_tests.py

Commit:
fix: undo draft

Commit:
fix: pre-commit

Commit:
fix: pylint

Commit:
fix: pylint additions

Commit:
fix: pre-commit #532

Commit:
fix: fixesss linting and precommit

Commit:
fix: fix typing

Commit:
fix: bed merge of feature_flags

Commit:
chore: remove unused import

Commit:
chore: remove experiment tox env

Commit:
chore: remove unused test config

Commit:
refactor: remove skipIf and set feature flag in the tests in beginning and end of relevant tests

Commit:
Revert "refactor: remove skipIf and set feature flag in the tests in beginning and end of relevant tests"

This reverts commit bf1b80cd

Commit:
test: add github action for the feature

Commit:
fix: name the github action

Commit:
fix: make action work the right SUPERSET_CONFIG module path

Commit:
fix: bad yaml syntax

Commit:
fix: missed changes

Commit:
Merge remote-tracking branch 'upstream/master' into feature/dash_level_access/dashboard_by_id_endpoints

* upstream/master: (29 commits)
  fix(presto): default unknown types to string type (#10753)
  feat(row-level-security): add base filter type and filter grouping (#10946)
  docs: add gallery screenshot & link in README (#10988)
  docs: add a "Gallery" page (#10968)
  build: add PR lint action (#10990)
  adding filters back that caused issues (#10989)
  chore: selectors refactor in SQLLab test suite (Cypress) (#10944)
  ESLint: Remove ts-ignore comments (#10933)
  style: fix checkbox color (#10970)
  fix: changed disabled rules in datasets module (#10979)
  fix: update the time filter for 'Last Year' option in explore (#10829)
  fix: use nullpool even for user lookup in the celery (#10938)
  Allow empty observations in alerting (#10939)
  fix: re-enabling several globally disabled lint rules (#10957)
  fix: setting specific exceptions common/query_context.py (#10942)
  Pylint disabled rule `pointless-string-statement` is not raising warining anymore - removing (#10975)
  fix: pylint disabled rules in dashboard/api.py (#10976)
  fix: removed disabled lint rule `too-many-locals` in connectors/base/models.py (#10958)
  ESLint: Re-enable rule no-access-state-in-setstate (#10870)
  fix: simply is_adhoc_metric (#10964)
  ...

Commit:
refactor: remove all unrelated change to manager.py to remove overhead of this PR

Commit:
chore: undo unrelevant change, moved to a separated PR

Commit:
Merge remote-tracking branch 'upstream/master' into feat/add_all_dashboards_permissions

* upstream/master: (46 commits)
  fix: surface connection error messages on the client (#11077)
  fix(jest): using UTC mock date (#11079)
  removing unused component (#11072)
  changing to the correct hex color (#11073)
  style: remove unecessary padding (#11071)
  fix: database list checkboxes (#11068)
  feat: adding all icons from the design system to the codebase (#11033)
  fix: sql lab autocomplete width (#11063)
  clickable labels have outlines, storybook shows them (#11034)
  fixed routes for customer in docs (#11052)
  Revert "style: fix checkbox color (#10970)" (#11051)
  feat: add "created by" to dashboard CRUD view (#11030)
  Changed `tags.py` and `helpers.py` in `models` module: removed disabled pylint rule `unused_import`, changed unused arguments to private and removed disabled rule `unused-argument. Removed redundant rules.` (#11037)
  chore: updated lint rules in models module (#11036)
  Removed disable global pytlint rule `standarderror-builtin` which isn't appearing for Python3 (#11038)
  Enabled argument-differ for bulk_delete (#11039)
  Enabled no-self-use pylint rule in security. Formatter (#11041)
  Changed variable name from capitals to lowercase and changed lint rule (#11044)
  Revert "ESLint: Re-enable rule default-props-match-prop-types (#10868)" (#11050)
  feat(saved_queries): add custom api filter for all string & text fields (#11031)
  ...

# Conflicts:
#	superset/config.py
#	tests/dashboards/api_tests.py

Commit:
Merge remote-tracking branch 'upstream/master' into feat/add_all_dashboards_permissions

* upstream/master: (46 commits)
  fix: surface connection error messages on the client (#11077)
  fix(jest): using UTC mock date (#11079)
  removing unused component (#11072)
  changing to the correct hex color (#11073)
  style: remove unecessary padding (#11071)
  fix: database list checkboxes (#11068)
  feat: adding all icons from the design system to the codebase (#11033)
  fix: sql lab autocomplete width (#11063)
  clickable labels have outlines, storybook shows them (#11034)
  fixed routes for customer in docs (#11052)
  Revert "style: fix checkbox color (#10970)" (#11051)
  feat: add "created by" to dashboard CRUD view (#11030)
  Changed `tags.py` and `helpers.py` in `models` module: removed disabled pylint rule `unused_import`, changed unused arguments to private and removed disabled rule `unused-argument. Removed redundant rules.` (#11037)
  chore: updated lint rules in models module (#11036)
  Removed disable global pytlint rule `standarderror-builtin` which isn't appearing for Python3 (#11038)
  Enabled argument-differ for bulk_delete (#11039)
  Enabled no-self-use pylint rule in security. Formatter (#11041)
  Changed variable name from capitals to lowercase and changed lint rule (#11044)
  Revert "ESLint: Re-enable rule default-props-match-prop-types (#10868)" (#11050)
  feat(saved_queries): add custom api filter for all string & text fields (#11031)
  ...

# Conflicts:
#	superset/config.py
#	tests/dashboards/api_tests.py

Commit:
Merge remote-tracking branch 'ofekNIelsen/feat/add_all_dashboards_permissions' into feat/add_all_dashboards_permissions

* ofekNIelsen/feat/add_all_dashboards_permissions:

# Conflicts:
#	tests/dashboards/api_tests.py
#	tests/dashboards/dashboard_test_utils.py

amitNielsen commented on superset/config.py:
```diff
@@ -308,6 +308,7 @@ def _try_json_readsha(  # pylint: disable=unused-argument
     "LISTVIEWS_DEFAULT_CARD_VIEW": False,
+    "DASHBOARD_LEVEL_ACCESS": False,
 }
 
```
move DASHBOARD_LEVEL_MODE=[""] as a config
